reference-implementation/lib/abstract-ops/readable-streams.js

```diff
@@ -422,6 +451,230 @@ function ReadableStreamTee(stream, cloneForBranch2) {
+  assert(ReadableStream.isImpl(stream));
+  assert(ReadableByteStreamController.isImpl(stream._controller));
+
+  let reader = AcquireReadableStreamDefaultReader(stream);
+  let reading = false;
+  let canceled1 = false;
+  let canceled2 = false;
+  let reason1;
+  let reason2;
+  let branch1;
+  let branch2;
+
+  const cancelPromise = newPromise();
+
+  function forwardReaderError(thisReader) {
+    uponRejection(thisReader._closedPromise, r => {
+      if (reader !== thisReader) {
+        return;
+      }
+      ReadableByteStreamControllerError(branch1._controller, r);
+      ReadableByteStreamControllerError(branch2._controller, r);
+      if (canceled1 === false || canceled2 === false) {
+        resolvePromise(cancelPromise, undefined);
+      }
+    });
+  }
+
+  function pullWithDefaultReader() {
+    if (ReadableStreamBYOBReader.isImpl(reader)) {
+      assert(reader._readIntoRequests.length === 0);
+      ReadableStreamReaderGenericRelease(reader);
+
+      reader = AcquireReadableStreamDefaultReader(stream);
+      forwardReaderError(reader);
+    }
+
+    const readRequest = {
+      chunkSteps: chunk => {
+        // This needs to be delayed a microtask because it takes at least a microtask to detect errors (using
+        // reader._closedPromise below), and we want errors in stream to error both branches immediately. We cannot let
+        // successful synchronously-available reads get ahead of asynchronously-available errors.
+        queueMicrotask(() => {
+          reading = false;
+
+          const chunk1 = chunk;
+          let chunk2 = chunk;
+          if (canceled1 === false && canceled2 === false) {
+            chunk2 = CloneArrayBufferView(chunk);
+          }
+
+          if (canceled1 === false) {
+            ReadableByteStreamControllerEnqueue(branch1._controller, chunk1);
+          }
+          if (canceled2 === false) {
+            ReadableByteStreamControllerEnqueue(branch2._controller, chunk2);
+          }
+        });
+      },
+      closeSteps: () => {
+        reading = false;
+        if (canceled1 === false) {
+          ReadableByteStreamControllerClose(branch1._controller);
+        }
+        if (canceled2 === false) {
+          ReadableByteStreamControllerClose(branch2._controller);
+        }
+        if (branch1._controller._pendingPullIntos.length > 0) {
+          ReadableByteStreamControllerRespond(branch1._controller, 0);
+        }
+        if (branch2._controller._pendingPullIntos.length > 0) {
+          ReadableByteStreamControllerRespond(branch2._controller, 0);
+        }
+        resolvePromise(cancelPromise, undefined);
+      },
+      errorSteps: () => {
+        reading = false;
+      }
+    };
+    ReadableStreamDefaultReaderRead(reader, readRequest);
+  }
+
+  function pullWithBYOBReader(view, forBranch2) {
+    if (ReadableStreamDefaultReader.isImpl(reader)) {
+      assert(reader._readRequests.length === 0);
+      ReadableStreamReaderGenericRelease(reader);
```

Comment:
Wow, so this switches off readers according to which each branch is doing? Very cool!

Reply:
Yup! I've done something similar before in [web-streams-adapter](https://github.com/MattiasBuelens/web-streams-adapter/blob/v0.1.0-alpha.5/src/readable-wrapper.ts#L235-L257), so that's where I got the idea. 😁